File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.cc
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>
#include <string.h>

using namespace std;
#define FASTIO ios::sync_with_stdio(false); cin.tie(0); cout.tie(0)

const int INF = 1987654321;

int n, m;
int arr[1001][1001];
int visited[1001][1001];
pair<int, int> start;

int dy[4] = {0, 1, 0, -1};
int dx[4] = {1, 0, -1, 0};

bool inRange(int y, int x) {
    return 0 <= y && y < n && 0 <= x && x < m;
}

void bfs() {
    queue<pair<int, int>> q;
    q.push(start);
    visited[start.first][start.second] = 1;

    while(!q.empty()) {
        pair<int, int> cur = q.front();
        q.pop();

        int y = cur.first, x = cur.second;
        for(int dir = 0; dir < 4; dir++) {
            int ny = y + dy[dir], nx = x + dx[dir];
            if(!inRange(ny, nx) || visited[ny][nx] || arr[ny][nx] == 0) continue;
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny, nx});
        }
    }
}

int main(void) {
    FASTIO;

    cin >> n >> m;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> arr[i][j];
            if(arr[i][j] == 2)
                start = {i, j};
        }
    }
    bfs();
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(arr[i][j] == 0)
                cout << "0 ";
            else if(!visited[i][j])
                cout << "-1 ";
            else
                cout << visited[i][j] - 1 << ' ';
        }
        cout << '\n';
    }
    
    return 0;
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.io.*;

public class Main {

    static class Pair {
        int y, x;
        public Pair(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
    
    static int n, m;
    static int[][] board = new int[1001][1001];
    static Pair start;
    static int[] dy = {0, 1, 0, -1};
    static int[] dx = {1, 0, -1, 0};
    static int[][] visited = new int[1001][1001];

    static boolean inRange(int y, int x) {
        return 0 <= y && y < n && 0 <= x && x < m;
    }

    static void bfs() {
        Queue<Pair> q = new LinkedList<>();
        visited[start.y][start.x] = 1;
        q.offer(start);

        while(!q.isEmpty()) {
            Pair here = q.peek();
            q.poll();

            int y = here.y, x = here.x;
            for(int dir = 0; dir < 4; dir++) {
                int ny = y + dy[dir], nx = x + dx[dir];
                if(!inRange(ny, nx) || board[ny][nx] != 1 || visited[ny][nx] != 0) continue;
                visited[ny][nx] = visited[y][x] + 1;
                q.offer(new Pair(ny, nx));
            } 
        }
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        m = sc.nextInt();
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                board[i][j] = sc.nextInt();
                if(board[i][j] == 2) {
                    start = new Pair(i, j);
                }
            }
        }
        bfs();
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(visited[i][j] == 0 && board[i][j] == 1)
                    System.out.print("-1 ");
                else if(visited[i][j] == 0)
                    System.out.print("0 ");
                else 
                    System.out.print(visited[i][j] - 1 + " ");
            }
            System.out.println();
        }
    }
}
